tracing-attributes/src/lib.rs

```diff
@@ -835,127 +968,158 @@ mod kw {
+    self_type: Option<syn::TypePath>,
+}
+
 // Get the AST of the inner function we need to hook, if it was generated
 // by async-trait.
 // When we are given a function annotated by async-trait, that function
 // is only a placeholder that returns a pinned future containing the
 // user logic, and it is that pinned future that needs to be instrumented.
 // Were we to instrument its parent, we would only collect information
 // regarding the allocation of that future, and not its own span of execution.
-// So we inspect the block of the function to find if it matches the pattern
-// `async fn foo<...>(...) {...}; Box::pin(foo<...>(...))` and we return
-// the name `foo` if that is the case. 'gen_body' will then be able
-// to use that information to instrument the proper function.
+// Depending on the version of async-trait, we inspect the block of the function
+// to find if it matches the pattern
+// `async fn foo<...>(...) {...}; Box::pin(foo<...>(...))` (old version), or if
+// it matches `Box::pin(async move { ... }) (new version). We the return the
+// statement that must be instrumented, along with some other informations.
+// 'gen_body' will then be able to use that information to instrument the
+// proper function/future.
 // (this follows the approach suggested in
 // https://github.com/dtolnay/async-trait/issues/45#issuecomment-571245673)
-fn get_async_trait_function(block: &Block, block_is_async: bool) -> Option<&ItemFn> {
+fn get_async_trait_info(block: &Block, block_is_async: bool) -> Option<AsyncTraitInfo<'_>> {
     // are we in an async context? If yes, this isn't a async_trait-like pattern
     if block_is_async {
         return None;
     }
 
     // list of async functions declared inside the block
-    let mut inside_funs = Vec::new();
-    // last expression declared in the block (it determines the return
-    // value of the block, so that if we are working on a function
-    // whose `trait` or `impl` declaration is annotated by async_trait,
-    // this is quite likely the point where the future is pinned)
-    let mut last_expr = None;
-
-    // obtain the list of direct internal functions and the last
-    // expression of the block
-    for stmt in &block.stmts {
-        if let Stmt::Item(Item::Fn(fun)) = &stmt {
-            // is the function declared as async? If so, this is a good
-            // candidate, let's keep it in hand
-            if fun.sig.asyncness.is_some() {
-                inside_funs.push(fun);
+    let inside_funs: Vec<(&Stmt, &ItemFn)> = block
+        .stmts
+        .iter()
+        .filter_map(|stmt| {
+            if let Stmt::Item(Item::Fn(fun)) = &stmt {
+                // If the function is async, this is a candidate
+                if fun.sig.asyncness.is_some() {
+                    return Some((stmt, fun));
+                }
             }
-        } else if let Stmt::Expr(e) = &stmt {
-            last_expr = Some(e);
+            None
+        })
+        .collect();
```

Comment:
is there a reason this has to be a vector? AFAICT, we only iterate over it in one place, it seems like we could just remove the `.collect()` and leave it as an iterator?